### ISSUE TITLE
knitr 1.45 seems to have broken the collectcode chunk option

### ISSUE THREAD

**Issue by remlapmot** [Hemken/Statamarkdown#38]
Hi Doug

knitr 1.45 was released on 30<sup>th</sup> October.

I think I notice that it has broken the collectcode chunk option.

I have put a demo at <https://github.com/remlapmot/collectcode-error> (the error is still present under Yihui's latest version on GitHub, but I didn't report this in the knitr repo yet).

* Correct output under knitr 1.44 <https://htmlpreview.github.io/?https://github.com/remlapmot/collectcode-error/main/test-knitr-1.44.html>
* Incorrect output under knitr 1.45 <https://htmlpreview.github.io/?https://github.com/remlapmot/collectcode-error/main/test-knitr-1.45.html>
* Incorrect output under knitr devel <https://htmlpreview.github.io/?https://github.com/remlapmot/collectcode-error/main/test-knitr-devel.html>

I haven't traced which change in knitr 1.45 causes the error but I know the differences between the knitr 1.44 and 1.45 can be viewed at

<https://github.com/yihui/knitr/compare/v1.44...v1.45>

If we can work out which was the breaking change maybe we should report to Yihui?

All the best
  Tom

**Comment by Hemken** [comment]
I’m out of town, traveling this week.

You could put a query at the beginning of the second Stata code chunk to see if the profile.do file exists at that point.  The error message suggests that auto.dta wasn’t even loaded.
________________________________
From: Tom Palmer ***@***.***>
Sent: Monday, November 6, 2023 8:49:00 AM
To: Hemken/Statamarkdown ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [Hemken/Statamarkdown] knitr 1.45 seems to have broken the collectcode chunk option (Issue #38)


Hi Doug

knitr 1.45 was released on 30th October.

I think I notice that it has broken the collectcode chunk option.

I have put a demo at https://github.com/remlapmot/collectcode-error (the error is still present under Yihui's latest version on GitHub, but I didn't report this in the knitr repo yet).

  *   Correct output under knitr 1.44 https://htmlpreview.github.io/?https://github.com/remlapmot/collectcode-error/main/test-knitr-1.44.html
  *   Incorrect output under knitr 1.45 https://htmlpreview.github.io/?https://github.com/remlapmot/collectcode-error/main/test-knitr-1.45.html
  *   Incorrect output under knitr devel https://htmlpreview.github.io/?https://github.com/remlapmot/collectcode-error/main/test-knitr-devel.html

I haven't traced which change in knitr 1.45 causes the error but I know the differences between the knitr 1.44 and 1.45 can be viewed at

***@***.***<https://github.com/yihui/knitr/compare/v1.44...v1.45>

If we can work out which was the breaking change maybe we should report to Yihui?

All the best
Tom

**Comment by remlapmot** [comment]
Interesting, under 1.45 profile.do isn't present after the first chunk. I'm not sure if it survives the first chunk.

**Comment by Hemken** [comment]
So I doubt it’s the concordance.R.  So which of these is it:  output.R, parser.R, utils.R?

--
Doug Hemken
SSCC statistical consultant
4226i Social Science Bldg.

***@***.******@***.***>
608-262-4327

**Comment by remlapmot** [comment]
I have tested by making the changes in the commit file by file and find that it's the changes in output.R - but I don't know which of the changes in that file causes the problem.

**Comment by Hemken** [comment]
Great!  I’ve just been setting up a test environment for myself, and that saves me several steps!

--
Doug Hemken
SSCC statistical consultant
4226i Social Science Bldg.

***@***.******@***.***>
608-262-4327

**Comment by Hemken** [comment]
Somehow those changes, which appear to be in a error handler (?) prevent the profile.do from being written (or from being written in the correct location?)

--
Doug Hemken
Statistical consultant
Social Science Computing Cooperative
Univ. of Wisc. – Madison

4226I Social Science Bldg.

***@***.***
608-262-4327

To make a consulting appointment send me an email, or use the on-line scheduler:
https://doodle.com/mm/doughemken/book-a-time

**Comment by remlapmot** [comment]
My only idea at the moment is whether the number in the `sys.frame()` call might need to be amended - maybe the new code has caused the numbers to change?

https://github.com/Hemken/Statamarkdown/blob/1e52b2370b623b7fdf20d87cb1068f9fa34cd216/R/stata_collectcode.r#L44

**Comment by remlapmot** [comment]
I amended that `-9` to a `-10` and collectcode seems to work again.

So we could set it conditionally on the knitr version number, i.e.,

```r
        if (packageVersion('knitr') < '1.45') {
            do.call("on.exit",
                list(quote(unlink("profile.do")), add=TRUE),
                envir = sys.frame(-9))
        }
        if (packageVersion('knitr') >= '1.45') {
            do.call("on.exit",
                list(quote(unlink("profile.do")), add=TRUE),
                envir = sys.frame(-10))
        }
```

(Also it's not clear to me whether we additionally need to amend the `sys.frame(-9)` in the if statement below this; currently on line 44, fot knitr 1.45)

**Comment by Hemken** [comment]
Brilliant!  (This is one of the two most obscure-to-me parts of the code I wrote!)

--
Doug Hemken
SSCC statistical consultant
4226i Social Science Bldg.

***@***.******@***.***>
608-262-4327

**Comment by Hemken** [comment]
Yes, those would both need to be changed.

Thanks for the code, I’ll implement it.

--
Doug Hemken
SSCC statistical consultant
4226i Social Science Bldg.

***@***.******@***.***>
608-262-4327

**Comment by remlapmot** [comment]
Great, thanks Doug.

It must be that the added `handle_error()` function call (or maybe the `knit_concord$get('block')`) bumps the sys.frame() number.

Do you think it is worth informing Yihui about this, so that he's aware of the issue in case any other knitr language engines have a similar issue?

**Comment by remlapmot** [comment]
Nb. I think I should have put the `utils::` package qualifier on the `packageVersion()` calls otherwise you'll get an R CMD check note, i.e., `utils::packageVersion('knitr')`

**Comment by Hemken** [comment]
It is my impression that this is becoming standard style across a number of programming languages.

--
Doug Hemken
SSCC statistical consultant
4226i Social Science Bldg.

***@***.******@***.***>
608-262-4327

**Comment by Hemken** [comment]
It will affect SASmarkdown, too.  But I’m not aware of many other language engines that work this way.

--
Doug Hemken
SSCC statistical consultant
4226i Social Science Bldg.

***@***.******@***.***>
608-262-4327

**Comment by remlapmot** [comment]
Ok, that's fine, I won't file an issue in the knitr repo.

**Comment by Hemken** [comment]
Just pushed a fix to Github.  Not sure when exactly, but I’ll run my usual battery of tests (both Windows and Linux) and then submit to CRAN.

--
Doug Hemken
SSCC statistical consultant
4226i Social Science Bldg.

***@***.******@***.***>
608-262-4327

**Comment by remlapmot** [comment]
That's great Doug

I have run my little test using your new version and knitr versions 1.43, 1.44, 1.45, and the latest knitr on GitHub and everything seems to be working well as before.

I'll try and keep it in my mind to run a test when I see new versions of knitr released.

**Comment by Hemken** [comment]
Fixed with Release 0.9.2
